### PR TITLE
[BEAM-1491]Identify HADOOP_CONF_DIR(or YARN_CONF_DIR) environment variables

### PR DESCRIPTION
**Pull request by 397090770**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by 397090770** [comment]
Hi @lukecwik, can you please review my PR? thank you.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java [hunk 1]
```diff
@@ -39,11 +45,37 @@
...
+      // Find default configuration when HADOOP_CONF_DIR or YARN_CONF_DIR is set.
+      Configuration conf = new Configuration(false);
+      List<String> hadoopEnvList = Lists.newArrayList("HADOOP_CONF_DIR", "YARN_CONF_DIR");
+      for (String env : hadoopEnvList) {
+        String hadoopConfPath = System.getenv(env);
+        if (!Strings.isNullOrEmpty(hadoopConfPath) && new File(hadoopConfPath).exists()) {
+
+          // We just need to load both core-site.xml and hdfs-site.xml to determine the
+          // default fs path and the hdfs configuration
+          if (new File(hadoopConfPath + "/core-site.xml").exists()) {
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java]: Please use the two argument constructor for File/Path here and below so we aren't assuming how path resolution works ("/" is common but not for every file system)

new File(hadoopConfPath, "core-site.xml");
new Path(hadoopConfPath, "core-site.xml");

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java [hunk 2]
```diff
@@ -39,11 +45,37 @@
...
+            conf.addResource(new Path(hadoopConfPath + "/core-site.xml"));
+
+            if (LOG.isDebugEnabled()) {
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java]: Please use parameterized messages
```
if (LOG.isDebugEnabled()) {
  LOG.debug("Log " + x);
}
```
becomes
```
LOG.debug("Log {}", x);
```
See:
https://www.slf4j.org/faq.html#logging_performance

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java [hunk 3]
```diff
@@ -39,11 +45,37 @@
...
     @Override
-    public Configuration create(PipelineOptions options) {
-      // TODO: Find default configuration to use
-      return null;
+    public List<Configuration> create(PipelineOptions options) {
+      // Find default configuration when HADOOP_CONF_DIR or YARN_CONF_DIR is set.
+      Configuration conf = new Configuration(false);
+      List<String> hadoopEnvList = Lists.newArrayList("HADOOP_CONF_DIR", "YARN_CONF_DIR");
+      for (String env : hadoopEnvList) {
+        String hadoopConfPath = System.getenv(env);
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java]: Add unit tests.

Note that `System.getenv` isn't mockable so the best bet is to make a method on `ConfigurationLocator` called:
```
@VisibleForTesting
Map<String, String> getEnvironment() {
  return System.getenv(key);
}
```

See [GcpOptions](https://github.com/apache/beam/blob/master/sdks/java/extensions/gcp-core/src/main/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptions.java#L189) and [GcpOptionsTest](https://github.com/apache/beam/blob/master/sdks/java/extensions/gcp-core/src/test/java/org/apache/beam/sdk/extensions/gcp/options/GcpOptionsTest.java#L124) for an example of how this kind of interaction can be tested

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java [hunk 4]
```diff
@@ -39,11 +45,37 @@
...
+          if (new File(hadoopConfPath + "/hdfs-site.xml").exists()) {
+            conf.addResource(new Path(hadoopConfPath + "/hdfs-site.xml"));
+
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Adding " + hadoopConfPath + "/hdfs-site.xml to hadoop configuration");
+            }
+          }
+        }
+      }
+      return Lists.<Configuration>newArrayList(conf);
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java]: We shouldn't be returning a configuration if we didn't load one from one of the paths.

File: sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java [hunk 5]
```diff
@@ -39,11 +45,37 @@
...
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationLocator.class);
     @Override
-    public Configuration create(PipelineOptions options) {
-      // TODO: Find default configuration to use
-      return null;
+    public List<Configuration> create(PipelineOptions options) {
+      // Find default configuration when HADOOP_CONF_DIR or YARN_CONF_DIR is set.
+      Configuration conf = new Configuration(false);
+      List<String> hadoopEnvList = Lists.newArrayList("HADOOP_CONF_DIR", "YARN_CONF_DIR");
+      for (String env : hadoopEnvList) {
```
Comment (lukecwik) [inline sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java]: If we find a configuration in HADOOP_CONF_DIR and YARN_CONF_DIR, we should be returning them both separately and not having the YARN_CONF_DIR overwriting the properties found in HADOOP_CONF_DIR.

Also, ensure that we only load one configuration if both HADOOP_CONF_DIR and YARN_CONF_DIR point to the same location.

**Comment by lukecwik** [comment]
Sorry for the churn since the SDK is going through several changes towards the first stable release.

**Pull request closed without merging**